core-primitives/primitives-cache/src/primitives_cache.rs

```diff
@@ -0,0 +1,115 @@
+/*
+	Copyright 2021 Integritee AG and Supercomputing Systems AG
+
+	Licensed under the Apache License, Version 2.0 (the "License");
+	you may not use this file except in compliance with the License.
+	You may obtain a copy of the License at
+
+		http://www.apache.org/licenses/LICENSE-2.0
+
+	Unless required by applicable law or agreed to in writing, software
+	distributed under the License is distributed on an "AS IS" BASIS,
+	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+	See the License for the specific language governing permissions and
+	limitations under the License.
+
+*/
+
+#[cfg(all(not(feature = "std"), feature = "sgx"))]
+use std::sync::SgxRwLock as RwLock;
+#[cfg(all(not(feature = "std"), feature = "sgx"))]
+use std::sync::SgxRwLockWriteGuard as RwLockWriteGuard;
+
+#[cfg(feature = "std")]
+use std::sync::RwLock;
+#[cfg(feature = "std")]
+use std::sync::RwLockWriteGuard;
+
+use std::string::{String, ToString};
+
+use crate::{
+	error::{Error, Result},
+	GetPrimitives, MutatePrimitives, Primitives,
+};
+
+/// Local primitives cache.
+///
+/// Stores the primitives internally, protected by a RW lock for concurrent access.
+#[derive(Default)]
+pub struct PrimitivesCache {
+	primitives_lock: RwLock<Primitives>,
+}
+
+impl PrimitivesCache {
+	pub fn new(primitives_lock: RwLock<Primitives>) -> Self {
+		PrimitivesCache { primitives_lock }
+	}
+}
+
+impl MutatePrimitives for PrimitivesCache {
+	fn load_for_mutation(&self) -> Result<RwLockWriteGuard<'_, Primitives>> {
+		self.primitives_lock.write().map_err(|_| Error::LockPoisoning)
+	}
+}
+
+impl GetPrimitives for PrimitivesCache {
+	fn get_primitives(&self) -> Result<Primitives> {
+		let primitives_lock = self.primitives_lock.read().map_err(|_| Error::LockPoisoning)?;
+		let primitives = &*primitives_lock;
+		Ok(primitives.clone())
```

Comment:
I can't. For your suggestion I get the following error:
```
error[E0507]: cannot move out of dereference of `SgxRwLockReadGuard<'_, Primitives>`
  --> /integriTEE/worker/core-primitives/primitives-cache/src/primitives_cache.rs:60:6
   |
60 |         Ok(*primitives_lock)
   |            ^^^^^^^^^^^^^^^^ move occurs because value has type `Primitives`, which does not implement the `Copy` trait
```
Switched to return the ReadGuard instead.